corehq/apps/user_importer/importer.py

```diff
@@ -582,35 +579,35 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
                     # Passing use_primary_db=True because of https://dimagi-dev.atlassian.net/browse/ICDS-465
                     user.get_django_user(use_primary_db=True).check_password(password)
 
-                for group in domain_info.group_memoizer.by_user_id(user.user_id):
-                    if group.name not in group_names:
-                        group.remove_user(user)
+                error = commcare_user_importer.update_user_groups(domain_info, group_names)
+                if error:
+                    error_message = f"{user.user_id}: {error}"
+                    ret['errors'].append(_(error_message))
+                try:
+                    for domain_info in domain_info_by_domain.values():
+                        domain_info.group_memoizer.save_all()
+                except BulkSaveError as e:
+                    _error_message = (
+                        "Oops! We were not able to save some of your group changes. "
+                        "Please make sure no one else is editing your groups "
+                        "and try again."
+                    )
+                    logging.exception((
+                        'BulkSaveError saving groups. '
+                        'User saw error message "%s". Errors: %s'
+                    ) % (_error_message, e.errors))
+                    ret['errors'].append(_error_message)
+
+                commcare_user_importer.save_log()
```

Comment:
@Robert-Costello I am a bit concerned by the order of actions here. `commcare_user_importer.save_log()` is called right in the end, way after `user.save`.
My understanding is that there is a possibility that user gets updated but the log is not saved?
Suggesting if we could keep the log save where it is and just update the change_messages if group update is successful?
Also, even currently, we have logged group updates but they can fail at the bulk update step, leading to an incorrect logging.

Reply:
Ok, I've put `commcare_user_importer.save_log()` back with `user.save()`. `commcare_user_importer.save_log` now returns the record. I modified `update_user_groups` to make the group updates and return new change messages. Once bulk upload has succeeded I'm updating the record with the new change messages. Haven't had a chance to add automated tests yet, but it does appear to work when testing locally.